fraud/service.go

```diff
@@ -0,0 +1,147 @@
+package fraud
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"sync"
+	"time"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+)
+
+// service is responsible for validating and propagating Fraud Proofs.
+// It implements the Service interface.
+type service struct {
+	mu           sync.RWMutex
+	unmarshalers map[ProofType]ProofUnmarshaler
+
+	topics *topics
+	getter headerFetcher
+}
+
+func NewService(p *pubsub.PubSub, getter headerFetcher) Service {
+	return &service{
+		topics: &topics{
+			pubsub:       p,
+			pubSubTopics: make(map[ProofType]*pubsub.Topic),
+		},
+		unmarshalers: make(map[ProofType]ProofUnmarshaler),
+		getter:       getter,
+	}
+}
+
+func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
+	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
+	f.mu.Lock()
+	u, ok := f.unmarshalers[proofType]
```

Comment:
Also I think that serialisation, validator registration/(deregistration?), topic should have a tighter coupling than this as you cannot have a topic w/o serialisation and should not have topic without validator. Then such checks can be avoided.

Something like: 

```
// topic constructor should also take validator so that there is no need for a separate register validator step, this should happen in the construction of Topic
type Topic struct {
 topic *topic
 unmarshaler ProofUnmarshaler
}
```

```
	pubSubTopics map[ProofType]*Topic
```

Reply:
bc `topics` inside topics.go is essentially a topics manager, why can't it also manage the topic's serialisation?

Reply:
This will simplify a lot of the code as well IMO.